### ISSUE TITLE
Weight decay is not multiplied with learning rate in DecoupledWeightDecayExtension 

### ISSUE THREAD

**Issue by kelvinkoh0308** [tensorflow/addons#2282]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux
- TensorFlow version and how it was installed (source or binary): Tensorflow 2.2
- TensorFlow-Addons version and how it was installed (source or binary): 0.11.2
- Python version: 3.7
- Is GPU used? (yes/no): Yes

**Describe the bug**

Hi, I have some question about the source code of DecoupledWeightDecayExtension.py because I found out that the variable is substracted with **weight decay** only, not **learning_rate * weight_decay**. Please point me out if I was wrong. 

```
    def _decay_weights_op(self, var, apply_state=None):
        if not self._decay_var_list or var.ref() in self._decay_var_list:
            var_device, var_dtype = var.device, var.dtype.base_dtype
            coefficients = (apply_state or {}).get(
                (var_device, var_dtype)
            ) or self._fallback_apply_state(var_device, var_dtype)

            return var.assign_sub(coefficients["wd_t"] * var, self._use_locking)
        return tf.no_op()
```

**Comment by WindQAQ** [comment]
Hi @kelvinkoh0308, thanks for the question. After scrutinizing the paper, our implementation is very different from it. In the paper, the weight decay is applied to gradients and parameters, but we only apply on parameters. We should correct it actually. This can be done along with optimizer cleanup. https://github.com/tensorflow/addons/issues/2187

**Comment by bhack** [comment]
@WindQAQ are we going to break some model reproducibility with this change in https://github.com/tensorflow/models/search?q=adamw or are their using its own reimplementation?

**Comment by WindQAQ** [comment]
Let alone correct implementation. If we adopt the change of #2187, it will at least break serialization (and compatibility and reproducibility) I believe. I am afraid that something must be changed in tfmodels.

**Comment by WindQAQ** [comment]
Seems that the implementation is inherited from `tf.contrib`.

**Comment by bhack** [comment]
I don't see I don't find `DecoupledWeightDecayExtension` occurence in TF/model probably they have duplicated their own AdamW/SGDW

**Comment by kelvinkoh0308** [comment]
@WindQAQ hi, I don't think that the weight decay is applied to gradients in the paper.

According to the image below,  the highlighted purple block is the L2 regularization which is added to the gradients and the green color block is the weight decay multiplied with the learning rate and then subtracted from the parameter. We should only follow the implementation according to the green color block only because the purple color color is the incorrect way to add weight decay.
 
![image](https://user-images.githubusercontent.com/50705244/101847507-61f87600-3b8e-11eb-9657-b9e2b2148593.png)

Currently, I already switch to use rectified adam from tensorflow addons which seems to be implemented correctly the weight decay.

**Comment by WindQAQ** [comment]
> @WindQAQ hi, I don't think that the weight decay is applied to gradients in the paper.
> 
> According to the image below, the highlighted purple block is the L2 regularization which is added to the gradients and the green color block is the weight decay multiplied with the learning rate and then subtracted from the parameter. We should only follow the implementation according to the green color block only because the purple color color is the incorrect way to add weight decay.
> 
> ![image](https://user-images.githubusercontent.com/50705244/101847507-61f87600-3b8e-11eb-9657-b9e2b2148593.png)
> 
> Currently, I already switch to use rectified adam from tensorflow addons which seems to be implemented correctly the weight decay.

Thanks for correcting me! You're right. I didn't see the coloring there. But, `\eta` is not the learning rate right? `\alpha` in the paper is the learning rate.

**Comment by kelvinkoh0308** [comment]
Ya, you are correct, alpha is the learning rate. This means that the weight decay is not multiplied with the learning rate.

I will close this issue. Thank you very much.

**Comment by kelvinkoh0308** [comment]
@WindQAQ  sry but this raises me another question for the rectified adam. 

https://github.com/tensorflow/addons/blob/81529ff7dd246f7575338b8cfe65784b0cc8a502/tensorflow_addons/optimizers/rectified_adam.py#L226

After added with the weight decay, the variable is updated with **lr_t * var_t**, does this means that the weight decay term is multiplied with the learning rate?

**Comment by WindQAQ** [comment]
> @WindQAQ sry but this raises me another question for the rectified adam.
> 
> https://github.com/tensorflow/addons/blob/81529ff7dd246f7575338b8cfe65784b0cc8a502/tensorflow_addons/optimizers/rectified_adam.py#L223-L226
> 
> After added with the weight decay, the variable is updated with **lr_t * var_t**, does this means that the weight decay term is multiplied with the learning rate?

Hi @kelvinkoh0308, no worries. I didn't take a look at the paper, but seems that author's implementation is the same with ours.

https://github.com/LiyuanLucasLiu/RAdam/blob/master/radam/radam.py#L82-L92
https://pytorch.org/docs/stable/tensors.html#torch.Tensor.add_

**Comment by kelvinkoh0308** [comment]
@WindQAQ if this is the case then I don't think I will use weight scheduler in RAdam as in AdamW since in AdamW the weight decay term is not multiplied with the learning rate thus we have to manually decay the weight decay term. However in the case of Rectified Adam, the weight decay term is multiplied with learning rate which means that the if we decay our learning rate the weight decay term will also be decayed automatically.

But this also means the weight decay rate will be larger in AdamW compared to RAdam if we passed the same weight decay eg. 1e-3 to these two optimizers because in RAdam the weight decay term will be multiplied with the learning rate

I hope I am correct. Thanks for you help.

**Comment by WindQAQ** [comment]
Yes you are right. If you want to do the correct comparison, maybe you should use larger weight decay for RAdam as what you suggested.

**Comment by kelvinkoh0308** [comment]
@WindQAQ thanks
Should they be standardized whether to multiplied with the learning rate in tensorflow_addons?

As this might lead to weird result for both optimizer if the users are not aware that one weight decay rate is larger than another optimizer because the documentation in tensorflow_addons does not mention this.

**Comment by WindQAQ** [comment]
> @WindQAQ thanks
> Should they be standardized whether to multiplied with the learning rate in tensorflow_addons?
> 
> As this might lead to weird result for both optimizer if the users are not aware that one weight decay rate is larger than another optimizer because the documentation in tensorflow_addons does not mention this.

It's weird to me to change the behavior because both implementations follow the paper/original impl directly. However, it's more than welcome to have more documentation.

**Comment by kelvinkoh0308** [comment]
For your reference, the implementation in pytorch is multiplied with learning rate

https://github.com/pytorch/pytorch/blob/9b0ffb9fb3a3b1b47c2d39270adf4f170f6a2141/torch/optim/adamw.py#L73

**Comment by WindQAQ** [comment]
Seems that pytorch considers `\eta * \alpha` as learning rate. We could do the same thing by passing a real `\eta` as argument. What do you think about this?

**Comment by kelvinkoh0308** [comment]
Right now both AdamW and RAdam in tensorflow_addons both can accept callable weight decay scheduler

This is true for AdamW as right now we have to decay manually, but is not required in the case of RAdam.

I think pytorch decided to go for multiplying with the learning rate in AdamW is because they don't implement weight scheduler (I think?) 

personally I don't think that add an extra `\eta` argument is a good choice as this will confuse with the learning rate scheduler passed to the learning_rate argument and also have to change the structure of the current implementation. 

What I prefer is that both implementation should be multiplied with learning rate, so that the user can adjust their weight decay term according to their learning rate.

**Comment by WindQAQ** [comment]
It's still hard to convince me to use a wrong implementation in order to match two implementations (Though we are also different from the paper). However, as a open source project, let's mark this discussion needed. @bhack and @seanpmorgan what do you think? Thank you.

BTW, here's author's implementation. It uses `\eta = current_learning_rate / init_learning_rate` and `current_weight_decay = \eta * init_weight_decay` so that both weight decay and learning rate decay in the same fashion. Seems that none of us follows author's implementations 😆 
https://github.com/loshchil/AdamW-and-SGDW/blob/master/train.lua

Thanks for opening the discussion!

**Comment by kelvinkoh0308** [comment]
@WindQAQ  I think if following this `\eta = current_learning_rate / init_learning_rate` and `current_weight_decay = \eta * init_weight_decay`  then current implementation in tensorflow_addons are correct because the weight decay is multiplied with `\eta` only

**Comment by kelvinkoh0308** [comment]
If you think that it is a wrong implementation if we change AdamW to multiply with learning rate with weight decay, then I think better to add to the documentation to tell user that the weight decay term will be larger because it is not scaled down by learning rate.

**Comment by WindQAQ** [comment]
> If you think that it is a wrong implementation if we change AdamW to multiply with learning rate with weight decay, then I think better to add to the documentation to tell user that the weight decay term will be larger because it is not scaled down by learning rate.

Totally fine with more documentation!
